finetune/lora.py

```diff
@@ -202,6 +207,46 @@ def pad_right(x, pad_id):
     return x, y
 
 
+class LlamaDataset(Dataset):
+    def __init__(self, data: list):
+        self._data = data
+
+    def __len__(self):
+        return len(self._data)
+
+    def __getitem__(self, i: int):
+        input_ids = self._data[i]["input_ids"].type(torch.int64)
+        labels = self._data[i]["labels"].type(torch.int64)
+        return input_ids, labels
+
+
+def get_dataloader(
+    fabric: L.Fabric,
+    data: torch.Tensor,
+    micro_batch_size: int,
+    group_by_length: bool,
+):
+    from transformers.trainer_pt_utils import LengthGroupedSampler
```

Comment:
Let's copy just this class with proper attribution in the docstring (it's a very small class and transformers is a big dependency)